Automattic/jetpack
# Widget Visibility: add controls to widget block editor

Fixes https://github.com/Automattic/jetpack/issues/20057

#### Changes proposed in this Pull Request:
* Re-implement widget visibility for block based widgets
  * Widget visibility is an existing jetpack feature for old-style widget (non-block based)
  * While creating and editing block based widgets, you may set up visibility rules inside the "Advanced" section of the Gutenberg sidebar
  * While visiting the front-end, widgets will hide and show depending on the rules that have been added
  * Note: While old-style widgets only had visibility rules for each widget, this PR allows visibility rules to be set up for each block within each widget, allowing for complex setups

![2021-08-25_14-21](https://user-images.githubusercontent.com/937354/130851964-fc30301b-2d64-4913-adf7-2bb64572ac10.png)
![2021-08-23_09-37](https://user-images.githubusercontent.com/937354/130466373-88d160ef-2503-4083-abf1-34c018881a1f.png)
![2021-08-23_09-37_1](https://user-images.githubusercontent.com/937354/130466388-969de40e-7776-4021-8167-0622ff8bbff1.png)


#### Issues

* **(2) Performance monitoring.** mtias asks "Are we monitoring any performance changes this might introduce?"
  * We are not, and I'm currently not sure how to handle this. 
  * Is there prior art for doing this?


#### Issues (Fixed)

 * ~~`widget_conditions_data` has some nested arrays as options, that render fancy nested selects.~~ Fixed, however it isn't perfect.
   * I was able to make the header options unselectable by disabling them, but gutenberg does not have any components that use `<optgroup` elements. (which is what the legacy visbility section uses)
 * ~~* Major option "Taxonomy" should be hidden if there's not a non-default taxonomy~~
 * ~~user and role major options should be hidden on wpcom~~. Fixed.
 * ~~Does not actually control visibility on the front end.~~ Fixed.
 * ~~Javascript build~~
   * ~~I'd like to write in JSX, but I haven't found a way. So I'm using element.createElement.~~
     * ~~Well, I did see instant search, which uses its own preact+webpack+babel setup, but it looked so complicated that I would probably be spending days just on environment setup. [babel](https://github.com/Automattic/jetpack/blob/2d74cbdeb98d117af8422b569da6ef056627376e/projects/plugins/jetpack/modules/search/instant-search/babel.config.js#L1), [package.json and more](https://github.com/Automattic/jetpack/blob/2d74cbdeb98d117af8422b569da6ef056627376e/projects/plugins/jetpack/package.json#L30-L33)~~
   * ~~I am using ES6 features, but I think my code isn't being run by babel, so I might have to convert it to ES5 (this could be tricky with the immutable updates).~~
   * Implemented Jeremy's build process. Thanks!
 * ~~Technically, legacy widgets are also blocks. This means you can currently specify their visibility on two levels, the legacy level and the new block based one, which would lead to confusing results.~~ Fixed: These are no longer allowed to have controls.
 *  ~~Accessibility issues on the delete button~~
 *  ~~On this branch, the /wp-admin/widgets.php page loads with the footer collapsed. There is an uncaught `Exception { name: "NS_ERROR_FAILURE", message: "", result: 2147500037,` shown in the browser console that's related to this behavior.~~
    * Fixed here: e00a1f5cc. It was happening when we were running `setAttributes` on the `core/legacy-widget` and `core/widget-area` blocks, which don't need visibility settings.
 * ~~I cannot add attributes to these types of blocks: 'core/archives', 'core/latest-comments', 'core/latest-posts'. Most blocks let me add a custom attribute, but with these, they start crashing when I add an attribute to them. Why? And are there others that will crash that I haven't found yet?~~
   * fix is in eb6564329d1baa5dc5185e5c7bbf552ec6aae588 
* ~~**(1) Customizer Bug.** F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Advanced -> Add new rule.~~
  * Expect to see: A new rule added and you are still looking at the list of rules.
  * Actually see: A new rule added and you are back to looking at the block widget. You must: Three dots menu -> Show more settings -> Advanced -> Add new rule.
  * **This bug is non-unique to the code in this branch.** It's an existing bug. For example, try:   F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Set the "Font Size" select to Large.  Now you're brought back to the block widget in the same manner.
  * **This bug is caused by code in core WP.org.** Commenting out [these 8 lines](https://github.com/WordPress/wordpress-develop/blame/d4abd5ca4a500f7b704c59d7a03ac1c583fde6a3/src/js/_enqueues/wp/customize/controls.js#L7199-L7206) stops it from happening. They were added in [this commit](https://github.com/WordPress/wordpress-develop/commit/d2540cfa069c23adc2f214bb40a0bb51eee445d3). [Comment in trac thread](https://core.trac.wordpress.org/ticket/37661#comment:216)
    * **Even though code in WP.org is causing it, it does not happen on a standalone .org installation.** This is one of those tricky issues that is happening in an interplay between wp.com and wp.org code.
  * I don't think this bug should be solved in this PR, but elsewhere. Should it be a blocker for this feature?
  * Reported here: https://github.com/Automattic/wp-calypso/issues/55875
  * And on .org trac: https://core.trac.wordpress.org/ticket/54063
  * Created Gutenberg PR which I believe fixes this: https://github.com/WordPress/gutenberg/pull/34543
  * PR will be released in Gutenberg 11.5, scheduled for release on Sept. 15

#### Random Musings

 * If you can set visibility rules on any block, it feels like there are so many different places to set them, it could be a potential foot-gun.

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:
* `jetpack docker up`
* Make sure WP is installed and you can visit /wp-admin. You may need to use `jetpack docker install`, and possibly edit one of the env files such that your password isn't easily guessable if you decide to use a tunneling solution in the future.
* In another terminal, `jetpack watch plugins/jetpack`
  * You might also need to `jetpack build plugins/jetpack` first.
* Make sure this setting is turned on in jetpack settings.
![2021-08-17_17-46](https://user-images.githubusercontent.com/937354/129810581-2a3911d3-7670-4843-a08d-5707813ef5da.png)
* Visit http://localhost/wp-admin/widgets.php
* Add a block based widget here
* In any block, go to the advanced section of the gutenberg sidebar

![2021-08-17_17-39](https://user-images.githubusercontent.com/937354/129809985-92438675-d682-4463-9889-b163cef6ad54.png)
^ Here it is with no design (although, it might be challenging to fit in sidebar)

Compare to the legacy visibility which we are re-implementing:
![2021-08-17_17-40](https://user-images.githubusercontent.com/937354/129810046-2a7fa9f4-e81a-4442-8a5b-8379d6d5ea4f.png)
^ Legacy visibility, not in this PR

#### Testing on DotCom
* Sandbox a simple site
* Apply D65619-code
